quickwit/quickwit-serve/src/metrics.rs

```diff
@@ -17,24 +17,142 @@
-use quickwit_common::metrics::{new_counter, IntCounter};
+use quickwit_common::metrics::{
+    new_counter, new_counter_vec, new_histogram_vec, Histogram, HistogramVec, IntCounter,
+    IntCounterVec,
+};
+use tower_http::classify::{ServerErrorsAsFailures, SharedClassifier};
+use tower_http::trace::{
+    DefaultMakeSpan, DefaultOnBodyChunk, DefaultOnEos, OnFailure, OnRequest, OnResponse, TraceLayer,
+};
 
 pub struct RestMetrics {
-    pub http_requests_total: IntCounter,
+    pub http_requests_total: IntCounterVec<2>,
+    pub http_requests_error: IntCounter,
+    pub http_requests_duration_secs: HistogramVec<2>,
 }
 
 impl Default for RestMetrics {
     fn default() -> Self {
         RestMetrics {
-            http_requests_total: new_counter(
+            http_requests_total: new_counter_vec(
                 "http_requests_total",
                 "Total number of HTTP requests received",
                 "quickwit",
+                ["method", "path"],
+            ),
+            http_requests_error: new_counter(
+                "http_requests_error",
+                "Number of HTTP requests with errors",
+                "quickwit",
+            ),
+            http_requests_duration_secs: new_histogram_vec(
+                "http_requests_duration_secs",
+                "Number of seconds required to run the HTTP request",
+                "quickwit",
+                ["method", "path"],
             ),
         }
     }
 }
 
+pub type RestMetricsTraceLayer<B> = TraceLayer<
+    SharedClassifier<ServerErrorsAsFailures>,
+    DefaultMakeSpan,
+    RestMetricsRecorder<B>,
+    RestMetricsRecorder<B>,
+    DefaultOnBodyChunk,
+    DefaultOnEos,
+    RestMetricsRecorder<B>,
+>;
+
+/// Holds the state required for recording metrics on a given request.
+pub struct RestMetricsRecorder<B> {
+    pub histogram: Arc<Mutex<Option<Histogram>>>,
+    _phantom: PhantomData<B>,
+}
+
+impl<B> Clone for RestMetricsRecorder<B> {
+    fn clone(&self) -> Self {
+        Self {
+            histogram: self.histogram.clone(),
+            _phantom: self._phantom,
+        }
+    }
+}
+
+impl<B> RestMetricsRecorder<B> {
+    pub fn new() -> Self {
+        Self {
+            histogram: Arc::new(Mutex::new(None)),
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<B, FailureClass> OnFailure<FailureClass> for RestMetricsRecorder<B> {
+    fn on_failure(
+        &mut self,
+        _failure_classification: FailureClass,
+        latency: std::time::Duration,
+        _span: &tracing::Span,
+    ) {
+        SERVE_METRICS.http_requests_error.inc();
+        self.clone().record_latency(latency);
+    }
+}
+
+impl<B> RestMetricsRecorder<B> {
+    fn record_latency(self, latency: Duration) {
+        if let Some(mutex) = Arc::into_inner(self.histogram) {
+            if let Some(histogram) = mutex.into_inner().expect("Failed to unlock histogram") {
+                histogram.observe(latency.as_secs_f64());
+            }
+        }
```

Comment:
found about `into_inner` method (https://doc.rust-lang.org/std/sync/struct.Arc.html#method.into_inner)

Reply:
but looks like it's not working 😅